discovery-infra/test_infra/helper_classes/kube_helpers.py

```diff
@@ -0,0 +1,376 @@
+import logging
+import os
+import contextlib
+
+import waiting
+
+from typing import Optional, Union, Dict, Tuple, ContextManager
+
+from kubernetes.config import load_kube_config
+from kubernetes.config.kube_config import Configuration
+from kubernetes.client import ApiClient, CoreV1Api, CustomObjectsApi
+from kubernetes.client.rest import ApiException
+
+from tests.conftest import env_variables
+from test_infra.utils import get_random_name
+
+
+# silence kubernetes debug messages.
+logging.getLogger('kubernetes').setLevel(logging.INFO)
+
+logger = logging.getLogger(__name__)
+
+
+def _does_string_contain_value(s: Union[str, None]) -> bool:
+    return s and s != '""'
+
+
+class ObjectReference(dict):
+
+    def __init__(self, name: str, namespace: str):
+        dict.__init__(self, name=name, namespace=namespace)
+
+
+class _BaseKubeAPIResource(object):
+    _shared_api_client = []
+
+    def __init__(self, name: str, namespace: str):
+        self.__ref = ObjectReference(name=name, namespace=namespace)
+
+        self._set_api_client_if_needed()
+
+    @property
+    def ref(self) -> ObjectReference:
+        return self.__ref
+
+    @property
+    def api_client(self) -> ApiClient:
+        return self._shared_api_client[0]
+
+    @classmethod
+    def _set_api_client_if_needed(cls) -> None:
+        # reuse same client for all subclasses.
```

Comment:
I'm not sure it would work as a Singleton for parallel multi-thread tests. You should look for Borg pattern.

Reply:
I change it to get the client from outside (which makes this base class a little bit unnecessary, so I removed it)